lib/spack/spack/url_buildcache.py

```diff
@@ -0,0 +1,1262 @@
+        return URLBuildcacheEntry
+    else:
+        raise UnknownBuildcacheLayoutError(
+            f"Cannot create buildcache class for unknown layout version {layout_version}"
+        )
+
+
+def check_mirrors_for_layout():
+    """Check configured mirrors, warning about any that are missing layout.json"""
+    if not config.get("config:check_mirrors_on_startup"):
+        return
+
+    cache_class = get_url_buildcache_class()
+    for mirror in MirrorCollection(binary=True).values():
+        if not cache_class.check_layout_json_exists(mirror.fetch_url):
+            msg = (
+                f"Configured mirror {mirror.name} is missing layout.json and has either \n"
+                "    never been pushed or is of an old layout version. Consider running \n"
+                "    'spack buildcache migrate' or rebuilding the specs in this mirror."
+            )
+            tty.warn(msg)
+
+
+def validate_checksum(file_path, checksum_algorithm, expected_checksum) -> None:
+    """Compute the checksum of the given file and raise if invalid"""
+    local_checksum = spack.util.crypto.checksum(hash_fun_for_algo(checksum_algorithm), file_path)
+
+    if local_checksum != expected_checksum:
+        size, contents = fsys.filesummary(file_path)
+        raise spack.error.NoChecksumException(
+            file_path, size, contents, checksum_algorithm, expected_checksum, local_checksum
+        )
+
+
+def _get_compressor(compression: str, writable: io.IOBase) -> io.IOBase:
+    if compression == "gzip":
+        return gzip.GzipFile(filename="", mode="wb", compresslevel=6, mtime=0, fileobj=writable)
+    elif compression == "none":
+        return writable
+    else:
+        raise BuildcacheEntryError(f"Unknown compression type: {compression}")
+
+
+@contextmanager
+def compression_writer(output_path: str, compression: str, checksum_algo: str):
+    """Create and return a writer capable of writing compressed data. Available
+    options for compression are "gzip" or "none", checksum_algo is used to pick
+    the checksum algorithm used by the ChecksumWriter.
+
+    Yields a tuple containing:
+        io.IOBase: writer that can compress (or not) as it writes
+        ChecksumWriter: provides checksum and length of written data
+    """
+    with open(output_path, "wb") as writer, ChecksumWriter(
+        fileobj=writer, algorithm=hash_fun_for_algo(checksum_algo)
+    ) as checksum_writer, closing(
+        _get_compressor(compression, checksum_writer)
+    ) as compress_writer:
+        yield compress_writer, checksum_writer
+
+
+def compress_and_write_spec(
+    output_path: str, spec_dict: dict, checksum_algo: str
+) -> Tuple[str, int]:
+    """Compress the spec dict and write it to the given path
+
+    Return the checksum (using the given algorithm) and size on disk of the file
+    """
+    with compression_writer(output_path, "gzip", "sha256") as (compressor, checker):
+        compressor.write(json.dumps(spec_dict).encode("utf-8"))
```

Comment:
Avoid temporary str and bytes in memory; and do the usual separators=... to save bytes. For example:

```suggestion
    with compression_writer(output_path, "gzip", "sha256") as (f_bin, checker), io.TextIOWrapper(f_bin, encoding="utf-8") as f_txt:
        json.dump(spec_dict, f_txt, separators=(",", ":"))
```

Or make `compression_writer` accept an open mode -- but then you have to deal with Literal and @overload type of stuff to make mypy happy and it might be too many LOC.